### PR TITLE
Improve logic for getNumPins

### PR DESCRIPTION
**Pull request by mgjarrett**
## Description

#1096 

The logic for `Block.getNumPins()` will identify lots of different components as pins, whether or not they are indeed pins. If a component has `Flags.SHIELD`, for example, it will be considered a pin (no matter what the shape or intent of the actual component is).

It is a relatively safe bet to assume all pins are circular, so a filter on component shape is applied to weed out any non-pin components.

---

## Checklist

- [x] This PR has only one purpose or idea.
- [x] Tests have been added/updated to verify that the new/changed code works.

- [x] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [x] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [x] The [release notes](https://terrapower.github.io/armi/release/index.html) (location `doc/release/0.X.rst`) are up-to-date with any bug fixes or new features.
- [x] The documentation is still up-to-date in the `doc` folder.
- [x] The dependencies are still up-to-date in `setup.py`.

### CODE REVIEW

File: armi/reactor/tests/test_blocks.py [hunk 1]
```diff
@@ -1242,6 +1242,11 @@ def test_getNumPins(self):
         emptyBlock = blocks.HexBlock("empty")
         self.assertEqual(emptyBlock.getNumPins(), 0)
 
+        comp = basicShapes.Hexagon("hexagon", "HT9", 1, 1, 1)
```
Comment (john-science) [inline armi/reactor/tests/test_blocks.py:1245]: I love it!

You've added so much logic to `getNumPins()` I wonder if we can test it a bit more.

For instance, can we have one test where it returns a NON-zero pin count, because it has a circular shape somewhere in it?  Perhaps we could do one with a `HoledRectangle`, and prove that the hole in the rectangle?

https://github.com/terrapower/armi/blob/20ca6877c2e2221bc3e3521cf9e827cf2d4a4376/armi/reactor/components/complexShapes.py#L94

Comment (mgjarrett) [reply armi/reactor/tests/test_blocks.py]: I added some more tests. Let me know if you think there are other corner cases we should cover!

**Review by john-science — APPROVED** [review]
Looks great!
